File: src/lib/extractor/hljs-validator.ts
// CodeLooter — Multi-layer code validation using highlight.js
//
// highlight.js is a syntax highlighter with built-in language auto-detection.
// It assigns a "relevance" score to each text snippet — higher = more likely
// to be valid code in some language.
//
// We use it as a SCREENING LAYER:
// 1. Pattern matching extracts candidate code blocks (fast, regex)
// 2. highlight.js validates each block — is it really code? (syntax-aware)
// 3. If relevance >= threshold → keep as code. If < threshold → likely narrative.
//
// This catches false positives (narrative that looks like code) and false
// negatives (code that doesn't match regex but IS valid syntax).
//
// Size: highlight.js ~9MB (npm), zero dependencies, pure JavaScript.
// Fits Render free tier (512MB) and Vercel free tier (1024MB).

import hljs from "highlight.js";

// Relevance threshold for multi-line blocks.
// From testing: real code blocks score 8-30+, narrative scores 0-5.
// We use 6 as threshold — narrative rarely exceeds 5, code rarely below 8.
const BLOCK_RELEVANCE_THRESHOLD = 6;

// For single lines, highlight.js is unreliable (too many false positives).
// We use a HIGHER threshold and additional filters to reduce false positives.
const LINE_RELEVANCE_THRESHOLD = 5;

// Lines starting with these patterns are NOT code, even if hljs thinks so.
const NARRATIVE_PATTERNS = [
  /^\s*-\s/,              // bullet points: "- Author (Year). Title"
  /^\s*\d+\.\s/,          // numbered lists: "1. Lakukan analisis"
  /^\s*Referensi\s*:/i,   // "Referensi:"
  /^\s*Penugasan\s*:/i,   // "Penugasan:"
  /^\s*Interpretasi\s*:/i,// "Interpretasi:"
  /^\s*Interpretasi\s+Hasil/i, // "Interpretasi Hasil"
  /^\s*Output\s+yang\s+dihasilkan/i,
  /^\s*Output\s+yang\s+dihasilkan\s*:/i,
  /^\s*Penjelasan\s*:/i,
  /^\s*Analisis\s*:/i,
  /^\s*Kesimpulan\s*:/i,
  /^\s*Kasus\s+\d/i,      // "Kasus 1:" header
  /^\s*Contoh\s+\d/i,    // "Contoh 1:" header
  /^\s*Latihan\s+\d/i,    // "Latihan 1:" header
  /^\s*Kode\s+[Pp]enyelesaian/i, // "Kode Penyelesaian:"
  /^\s*Kode\s+penyelesaiain/i,    // typo variant
  /^\s*Kode\s*:/i,
  // Narrative with statistical terms (false positive prone)
  /menunjukkan\s+bahwa/i,  // "menunjukkan bahwa..."
  /karena\s+p.?value/i,   // "karena p-value..."
  /diterima/i,             // "H0 diterima"
  /ditolak/i,              // "H0 ditolak"
  /Artinya/i,              // "Artinya, ..."
  /R.?squared/i,           // "R-squared = 0.667"
  /Adjusted/i,             // "Adjusted R-squared"
  /signifikan/i,           // "signifikan secara statistik"
  /koefisien/i,            // "koefisien..."
  /variabel/i,             // "variabel..."
  /peubah/i,               // "peubah..."
  /hipotesis/i,            // "hipotesis..."
  /taraf/i,                // "taraf nyata 5%"
  /sampel/i,               // "sampel data..."
  /penelitian/i,           // "penelitian..."
  /mahasiswa/i,            // "mahasiswa..."
  /dosen/i,                // "dosen..."
  /university/i,
  /hubungan\s+asosiasi/i, // "hubungan asosiasi"
];

export interface ValidationResult {
  isCode: boolean;
  relevance: number;
  language: string | null;
  secondBestLanguage: string | null;
  secondBestRelevance: number;
}

// Validate whether a text block is code or narrative using syntax highlighting.
// Works best with multi-line blocks (more context = better detection).
export function validateCodeBlock(code: string): ValidationResult {
  const result = hljs.highlightAuto(code);
  const relevance = result.relevance || 0;
  const language = result.language || null;
  const secondBestLanguage = result.secondBest?.language || null;
  const secondBestRelevance = result.secondBest?.relevance || 0;

  // Use higher threshold for blocks (more context available)
  const threshold = code.split("\n").length > 3
    ? BLOCK_RELEVANCE_THRESHOLD
    : LINE_RELEVANCE_THRESHOLD;

  return {
    isCode: relevance >= threshold,
    relevance,
    language,
    secondBestLanguage,
    secondBestRelevance,
  };
}

// Validate a single line — useful for ambiguous lines that pattern matching
// couldn't classify. Returns relevance score.
export function validateLine(line: string): ValidationResult {
  const t = line.trim();

  // Pre-filter: narrative patterns are never code, regardless of hljs score.
  for (const pattern of NARRATIVE_PATTERNS) {
    if (pattern.test(t)) {
      return {
        isCode: false,
        relevance: 0,
        language: null,
        secondBestLanguage: null,
        secondBestRelevance: 0,
      };
    }
  }

  const result = hljs.highlightAuto(line);
  return {
    isCode: (result.relevance || 0) >= LINE_RELEVANCE_THRESHOLD,
    relevance: result.relevance || 0,
    language: result.language || null,
    secondBestLanguage: result.secondBest?.language || null,
    secondBestRelevance: result.secondBest?.relevance || 0,
  };
}

// Get the detected language for a code block.
// Returns null if highlight.js can't determine the language.
export function detectLanguageHLJS(code: string): string | null {
  const result = hljs.highlightAuto(code);
  return result.language || null;
}

File: backend/app/hljs_validator.py
"""highlight.js validation layer for Python backend.

Uses Pygments (pure Python, ~5MB) as the equivalent of highlight.js
for syntax validation. Pygments can guess the lexer (language) for a
code snippet and returns a relevance score.

This is Layer 2 of the multi-layer extraction:
1. Pattern matching (regex) — fast, catches ~80% of code
2. Pygments validation — syntax-aware, filters false positives
3. Pygments recovery — catches code that pattern matching missed
"""
import re
import logging
from typing import Dict, Optional, Tuple

logger = logging.getLogger(__name__)

# Relevance threshold for multi-line blocks
BLOCK_RELEVANCE_THRESHOLD = 6

# For single lines, Pygments is less reliable — use higher threshold
LINE_RELEVANCE_THRESHOLD = 5

# Lines starting with these patterns are NOT code
NARRATIVE_PATTERNS = [
    re.compile(r"^\s*-\s"),              # bullet points
    re.compile(r"^\s*\d+\.\s"),          # numbered lists
    re.compile(r"^\s*Referensi\s*:", re.IGNORECASE),
    re.compile(r"^\s*Penugasan\s*:", re.IGNORECASE),
    re.compile(r"^\s*Interpretasi\s*:", re.IGNORECASE),
    re.compile(r"^\s*Output\s+yang\s+dihasilkan", re.IGNORECASE),
    re.compile(r"^\s*Penjelasan\s*:", re.IGNORECASE),
    re.compile(r"^\s*Analisis\s*:", re.IGNORECASE),
    re.compile(r"^\s*Kesimpulan\s*:", re.IGNORECASE),
    re.compile(r"^\s*Kasus\s+\d", re.IGNORECASE),
    re.compile(r"^\s*Contoh\s+\d", re.IGNORECASE),
    re.compile(r"^\s*Latihan\s+\d", re.IGNORECASE),
]

# UPGRADED (Task 21): Patterns that are DEFINITELY narrative, even if they contain
# = or (). Used as a pre-filter before pygments relevance scoring — matches lines
# like "X-squared = 2.2222 menunjukkan bahwa..." that pygments would otherwise
# mis-classify as code.
# IMPORTANT: single-word Indonesian terms use \b word boundaries so they don't
# match variable identifiers like `koefisien_variasi` or `signifikan_level`.
STATISTICAL_NARRATIVE_PATTERNS = [
    re.compile(r"menunjukkan\s+bahwa", re.IGNORECASE),
    re.compile(r"karena\s+p.?value", re.IGNORECASE),
    re.compile(r"\bH0\b.*diterima", re.IGNORECASE),
    re.compile(r"\bH0\b.*ditolak", re.IGNORECASE),
    re.compile(r"\bartinya\b", re.IGNORECASE),
    re.compile(r"\bR.?squared\b", re.IGNORECASE),
    re.compile(r"\bAdjusted\b", re.IGNORECASE),
    re.compile(r"\bsignifikan\b(?!_)", re.IGNORECASE),
    re.compile(r"\bkoefisien\b(?!_)", re.IGNORECASE),
    re.compile(r"hubungan\s+asosiasi", re.IGNORECASE),
    re.compile(r"\bpenyimpangan\b(?!_)", re.IGNORECASE),
    re.compile(r"\bdiperkirakan\b", re.IGNORECASE),
    re.compile(r"\bmeningkat\b", re.IGNORECASE),
    re.compile(r"\bberkontribusi\b", re.IGNORECASE),
    re.compile(r"\bmengindikasikan\b", re.IGNORECASE),
    re.compile(r"\bvariabilitas\b(?!_)", re.IGNORECASE),
    re.compile(r"\bdijelaskan\b", re.IGNORECASE),
]

_pygments_available = None


def _check_pygments():
    """Check if pygments is available."""
    global _pygments_available
    if _pygments_available is not None:
        return _pygments_available
    try:
        from pygments.lexers import guess_lexer
        _pygments_available = True
    except ImportError:
        _pygments_available = False
        logger.warning("Pygments not installed. Validation layer disabled.")
    return _pygments_available


def _get_relevance(code: str) -> Tuple[int, Optional[str]]:
    """Get relevance score and language for a code block using Pygments.
    
    Pygments doesn't return a numeric relevance score like highlight.js,
    but we can estimate it by counting how many tokens the lexer produces
    vs how much text remains untokenized.
    """
    if not _check_pygments():
        return 0, None
    try:
        from pygments.lexers import guess_lexer
        from pygments.token import Token
        
        lexer = guess_lexer(code)
        tokens = list(lexer.get_tokens(code))
        
        # Count "meaningful" tokens (not whitespace, not error)
        meaningful = 0
        errors = 0
        for tok_type, tok_val in tokens:
            if tok_type in (Token.Text, Token.Text.Whitespace):
                continue
            if tok_type in (Token.Error,):
                errors += 1
                continue
            meaningful += 1
        
        # Relevance = meaningful tokens - errors
        # More tokens = more likely to be valid code
        relevance = meaningful - errors * 2
        
        # Get language name
        lang_name = lexer.name.lower() if hasattr(lexer, 'name') else str(lexer.__class__.__name__).lower()
        
        return relevance, lang_name
    except Exception as e:
        logger.debug("Pygments relevance failed: %s", e)
        return 0, None


def validate_code_block(code: str) -> Dict:
    """Validate whether a text block is code or narrative.
    
    Returns:
        {
            "is_code": bool,
            "relevance": int,
            "language": str or None,
        }
    """
    relevance, language = _get_relevance(code)
    
    threshold = BLOCK_RELEVANCE_THRESHOLD
    if code.count("\n") <= 3:
        threshold = LINE_RELEVANCE_THRESHOLD
    
    return {
        "is_code": relevance >= threshold,
        "relevance": relevance,
        "language": language,
    }


def validate_line(line: str) -> Dict:
    """Validate a single line — useful for ambiguous lines.

    UPGRADED (Task 21): also filters statistical-narrative patterns
    ("X-squared = ... menunjukkan bahwa ...") that pygments would otherwise
    mis-classify as code.
    """
    t = line.strip()
    
    # Pre-filter: narrative patterns are never code
    for pattern in NARRATIVE_PATTERNS:
        if pattern.search(t):
            return {"is_code": False, "relevance": 0, "language": None}

    # Pre-filter (Task 21): statistical narrative patterns are never code
    for pattern in STATISTICAL_NARRATIVE_PATTERNS:
        if pattern.search(t):
            return {"is_code": False, "relevance": 0, "language": None}
    
    relevance, language = _get_relevance(t)
    return {
        "is_code": relevance >= LINE_RELEVANCE_THRESHOLD,
        "relevance": relevance,
        "language": language,
    }


def detect_language_pygments(code: str) -> Optional[str]:
    """Get the detected language for a code block."""
    if not _check_pygments():
        return None
    try:
        from pygments.lexers import guess_lexer
        lexer = guess_lexer(code)
        return lexer.name.lower() if hasattr(lexer, 'name') else None
    except Exception:
        return None
